src/auth/strategies/oidc.strategy.ts

```diff
@@ -0,0 +1,141 @@
+    private usersService: UsersService,
+  ) {
+    const oidcConfig = configService.get<Record<string, unknown>>("oidc");
+    super({
+      client: client,
+      params: {
+        redirect_uri: oidcConfig?.callbackURL,
+        scope: oidcConfig?.scope,
+        authorization_endpoint: oidcConfig?.authorizationURL,
+      },
+      passReqToCallback: false,
+      usePKCE: false,
+    });
+    this.client = client;
+  }
+
+  async validate(tokenset: TokenSet): Promise<Omit<User, "password">> {
+    const userinfo: UserinfoResponse = await this.client.userinfo(tokenset);
+
+    const userProfile = this.parseUserInfo(userinfo);
+
+    const userFilter: FilterQuery<UserDocument> = {
+      $or: [
+        { username: `oidc.${userProfile.username}` },
+        { email: userProfile.email as string },
+      ],
+    };
+    const userExists = await this.usersService.userExists(userFilter);
+    if (!userExists) {
+      const createUser: CreateUserDto = {
+        username: `oidc.${userProfile.username}`,
+        email: userProfile.email as string,
+      };
+
+      const user = await this.usersService.create(createUser);
+      if (!user) {
+        throw new InternalServerErrorException(
+          "Could not create User from OIDC response.",
+        );
+      }
+
+      const createUserIdentity: CreateUserIdentityDto = {
+        authScheme: "oidc",
+        created: new Date(),
+        credentials: {},
+        externalId: userProfile.id,
+        modified: new Date(),
+        profile: userProfile,
+        provider: "oidc",
+        userId: user._id,
+      };
+
+      await this.usersService.createUserIdentity(createUserIdentity);
+    }
+    const foundUser = await this.usersService.findOne(userFilter);
+    const jsonUser = JSON.parse(JSON.stringify(foundUser));
+    const { password, ...user } = jsonUser;
+    user.userId = user._id;
+
+    return user;
+  }
+
+  parseUserInfo(userinfo: UserinfoResponse) {
+    type OidcProfile = Profile & UserProfile;
+    const profile = {} as OidcProfile;
+    profile.id = userinfo.sub;
+    // Prior to OpenID Connect Basic Client Profile 1.0 - draft 22, the "sub"
+    // claim was named "user_id".  Many providers still use the old name, so
+    // fallback to that.
+    if (!profile.id) {
+      profile.id = userinfo.user_id as string;
+    }
+
+    profile.thumbnailPhoto = userinfo.thumbnailPhoto
+      ? "data:image/jpeg;base64," +
+        Buffer.from(userinfo.thumbnailPhoto as string, "binary").toString(
+          "base64",
+        )
+      : "error: no photo found";
+
+    if (userinfo.name) {
+      profile.displayName = userinfo.name;
+    }
+    if (userinfo.preferred_username) {
+      profile.username = userinfo.preferred_username;
+    }
+    if (userinfo.email) {
+      profile.emails = [{ value: userinfo.email }];
+      profile.email = userinfo.email;
+    }
+
+    profile.accessGroups = userinfo[
+      this.configService.get("oidc").accessGroups
```

Comment:
Does this mean that access groups are always hard-coded in configuration?

Reply:
only the key to retrieve the accessGroups values from the userInfo object (which is returned by the oidc when calling the userInfo endpoint):
Schematically, these are the steps: 

1. the code gets the oidc `.well-know` config
2. authentication is triggered through the oidc provider
3. when successful, oidc returns a token (jwt) 
4. token from 3 is used to call the oidc userInfo endpoint, which is contained in 1
5. 4 returns general user information, e.g. name, username etc and here this PR makes the assumption that one property of the userInfo is the accessGroups
6. line 136 extracts the accessGroups having specified the name of the userinfo accessGroups property in the config (e.g. at PSI we call them p-groups)